File: properties.cpp
#include "properties.h"
#include <stdexcept>
#include "CoolProp.h"

using CoolProp::PropsSI;

WaterProperties::WaterProperties(const std::string& fluid) : fluid_(fluid) {}

double WaterProperties::h_ps(double p, double s) const { return PropsSI("H","P",p,"S",s,fluid_); }
double WaterProperties::h_pt(double p, double T) const { return PropsSI("H","P",p,"T",T,fluid_); }
double WaterProperties::h_pq(double p, double q) const { return PropsSI("H","P",p,"Q",q,fluid_); }
double WaterProperties::s_pt(double p, double T) const { return PropsSI("S","P",p,"T",T,fluid_); }
double WaterProperties::s_ph(double p, double h) const { return PropsSI("S","P",p,"H",h,fluid_); }
double WaterProperties::s_pq(double p, double q) const { return PropsSI("S","P",p,"Q",q,fluid_); }
double WaterProperties::T_ph(double p, double h) const { return PropsSI("T","P",p,"H",h,fluid_); }
double WaterProperties::T_pq(double p, double q) const { return PropsSI("T","P",p,"Q",q,fluid_); }
double WaterProperties::saturationPressure(double T) const { return PropsSI("P","T",T,"Q",0,fluid_); }
double WaterProperties::saturationEnthalpy(double T, double q) const { return PropsSI("H","T",T,"Q",q,fluid_); }
double WaterProperties::saturationEntropy(double T, double q) const { return PropsSI("S","T",T,"Q",q,fluid_); }
double WaterProperties::criticalTemperature() const { return PropsSI("Tcrit",fluid_); }
double WaterProperties::tripleTemperature() const { return PropsSI("Ttriple",fluid_); }

File: properties.py
"""CoolProp wrapper.

Units:
    pressure: Pa
    temperature: K
    enthalpy: J/kg
    entropy: J/(kg K)
"""
from CoolProp.CoolProp import PropsSI

class WaterProperties:
    def __init__(self, fluid="Water"):
        self.fluid = fluid

    def h_ps(self, p, s):
        return PropsSI("H", "P", p, "S", s, self.fluid)

    def h_pt(self, p, T):
        return PropsSI("H", "P", p, "T", T, self.fluid)

    def h_pq(self, p, q):
        return PropsSI("H", "P", p, "Q", q, self.fluid)

    def s_pt(self, p, T):
        return PropsSI("S", "P", p, "T", T, self.fluid)

    def s_ph(self, p, h):
        return PropsSI("S", "P", p, "H", h, self.fluid)

    def s_pq(self, p, q):
        return PropsSI("S", "P", p, "Q", q, self.fluid)

    def T_ph(self, p, h):
        return PropsSI("T", "P", p, "H", h, self.fluid)

    def T_pq(self, p, q):
        return PropsSI("T", "P", p, "Q", q, self.fluid)

    def critical_temperature(self):
        return PropsSI("Tcrit", self.fluid)
